eip-0008.md

```diff
@@ -0,0 +1,479 @@
+else. For majority of scripts this is the case.
+
+ 2) The transaction sender must decide on the `costPrice` and `costLimit` values. The
+ `costLimit` parameter is used by the miners to reject over-expensive transactions and
+ should be greater or equal to `estCost` estimated during `tx` signing. The sender must
+ put at least `feeLimit = costLimit * costPrice` NERGs (NanoERGs) in the miner's fee
+ output box (see `feeOut` in the figure).
+The sender in addition puts the following values in the registers of `feeOut` box ([see
+footnote 2](#footnotes)):
+    - R4: a `Long` value of `costPrice` which is equal to the number of NanoERG to be
+    paid per unit of cost.
+    - R5: a `senderPk` public key belonging to the sender. The ownership is proved
+    during `tx` signing by spending at least one box protected by the `p2pk(senderPk)`
+    contract.
+    - R6: an interval of blockchain heights in which the transaction is valid, a pair
+    of two Int values.
+ 
+3) The transaction sender needs to convince miners that he has enough coins to pay for
+transaction fees up to `feeLimit`. He can do it by showing enough boxes protected by
+`p2pk(senderPk)` script corresponding to the public key stored in `feeOut.R5` register.
+When `tx` has predictable cost (i.e. all input scripts have predictable cost and `estCost
+== actCost`) then `feeLimit` is predictable and the sender puts enough inputs to cover it
+(see `inputs` and `collatInputs` in the figure). The `collatInputs` boxes are put at the
+tail of INPUTS collection, and the length of the tail is the smallest number of boxes to
+cover `feeLimit`.
+
+    However, when `actCost` is not predictable, which may happen when `tx` have different
+    costs depending on the actual block (i.e. scripts have conditions on `HEIGHT`,
+    `CONTEXT.headers`, etc), then the sender can choose a larger `costLimit` parameter to
+    guarantee the transaction is not rejected as over-costed. In this case `feeLimit` may
+    be much higher than `actCost * costPrice` and the sender will receive the `feeLimit -
+    actCost * costPrice` amount of refund (see `refundBox` in the figure).
+    
+    The total value of NERGs in `collatInputs` must be greater or equal to `feeLimit`. The
+    `change` output box is balancing the total amount in `collatInputs` with `feeLimit` (see
+    the figure). The `inSum` amount of `inputs` boxes should be balanced with the `outSum`
+    amount of `outputs` boxes. This balance should be independent of the balance of
+    `collatInputs`. This is necessary when the transaction is interpreted as the _charge
+    transaction_, in which case `inputs` are not spent and `outputs` are not created, so the
+    effect on the UTXO still remains balanced.
+
+    In an exceptional case of invalid transaction (due to the scripts returning false, or
+    `actCost > feeLimit`) the transaction will still be added to the block but will be
+    interpreted as a _charge transaction_ (see step 7 below).
```

Comment:
Also please note that "false" could mean invalid signature (unlike Ethereum where signature validation done separately). Unlikely including inputs with invalid signatures could be done securely.

Reply:
1) DataInputs also break the original UTXO semantics. Which UTXO semantics you mean?
2) > lso please note that "false" could mean invalid signature (unlike Ethereum where signature validation done separately).
I don't get the point. What is the problem exactly? (if you mean it is a problem)

Reply:
I would agree that this really does change the inherent model/semantics of UTXO system. I would err on the side of caution as something that major could incur serious problems down the line which we may not be able to see until they happen. (As we're changing from a properly formed tx accepted by a node **will** spend an input UTXO now to **may**)

Reply:
See the answer below. The attack you mention is prevented by those checks.

Reply:
> The most trivial attack is that malicious miners can always include pk("spender_key") boxes with invalid signatures (maybe, along with the spender-pk contracts) to get paid. could be fixed by complicating the protocol further (requiring valid signatures from pk("spender_key") boxes, but could be other vectors then you need to consider

On the other hand, if a transaction depends on the external environment (e.g. the hash of the previous block), It might fail only when miners packing the transaction. This might introduce some attack vectors?